### PR TITLE
Merge the constraint and state-read ASM and VM crates into `essential-asm`, `essential-vm`

### LINKED ISSUE #204: Unify ASM
With the changes being made to generalize over programs we don't need separate asm's.
The `Constraint` asm can be moved a group up and instead of calling the top level `StateRead` it should just be named something like `Ops`.

### LINKED ISSUE #205: Unify VM
The state read and constraint VM's can be unified now. I think this is basically as simple as pulling the constraint vm code into the state read vm crate and dedupping a few things.

### PR DESCRIPTION
**Pull request by mitchmindtree**
Edit: OK I think this is just about good to go!

## `Op` Refactor?

So far, the ASM enums still have the same grouping as they did before, but I'm wondering if it might be nice to also:

- flatten the `Constraint` group ops and
- Make the `StateRead` group only contain the `KeyRange*` ops, rather than act as a superset of all ops including the state read ops.

This means at the top `Op` level we would have the following groups: `Stack`, `Pred`, `Alu`, `Access`, `Crypto`, `ControlFlow`, `Memory`, `StateRead`.

cc @supiket @nfurfaro @mohammadfawaz @otrho thoughts? Either way, I might save this for a closely-followed PR as this one is already a lot to grok.

---

Related #172
Closes #204 
Closes #205

### CODE REVIEW

**Comment by mohammadfawaz** [comment]
Woot!

**Comment by mitchmindtree** [comment]
> Op Refactor?

I've decided to have a go at this as a part of this PR too.

File: crates/asm/src/lib.rs [hunk 1]
```diff
@@ -190,42 +193,31 @@ mod tests {
...
-            Stack::Push(0x1234567812345678).into(),
-            Stack::Swap.into(),
+            StateRead::KeyRange.into(),
+            StateRead::KeyRangeExtern.into(),
             Stack::Push(0x0F0F0F0F0F0F0F0F).into(),
-            Stack::Dup.into(),
-            Stack::Push(0x1234567812345678).into(),
         ];
         roundtrip(ops);
     }
```
Comment (supiket) [inline crates/asm/src/lib.rs:211]: Why is this removed?

**Review by supiket — COMMENTED** [review]
Echoing our IRL discussion: I agree with the changes planned to the ops in the way you described them in the issue. Seeing that the asm crate has a lot of changes related to these, I'm happy for those changes to land in this PR as well. Can be a followup PR if you want of course ^^

File: crates/vm/src/error.rs [hunk 2]
```diff
@@ -1,15 +1,31 @@
...
+/// Shorthand for a `Result` where the error type is an `OpSyncError`.
+pub type OpSyncResult<T> = Result<T, OpSyncError>;
+
+/// Shorthand for a `Result` where the error type is an `OpAsyncError`.
+pub type OpAsyncResult<T, E> = Result<T, OpAsyncError<E>>;
+
 /// Shorthand for a `Result` where the error type is a `CheckError`.
 pub type CheckResult<T> = Result<T, CheckError>;
 
+/// Shorthand for a `Result` where the error type is an `OpError`.
```
Comment (nfurfaro) [inline crates/vm/src/error.rs:26]:
```suggestion
/// Shorthand for a `Result` where the error type is a `ConstraintError`.
```

Comment (nfurfaro) [reply crates/vm/src/error.rs:26]: Not sure here... maybe OpError is correct?

**Review by nfurfaro — COMMENTED** [review]
A lot to grok, but looks good overall. Just a few small nits; will have to see what @supiket thinks as she has more context.

File: crates/asm/src/lib.rs (hunk 1 above)
Comment (mitchmindtree) [reply crates/asm/src/lib.rs]: Wow I have no idea, good catch! Re-added

**Comment by mitchmindtree** [comment]
@nfurfaro @supiket thanks for taking a look!

I've just finished the ASM refactor and some related renamings.

The `asm.yml` diff hasn't turned out very friendly, but fwiw it's mostly just indentation changes. Only the `Constraint` group has been removed, and the `KeyRange` ops have been moved under a dedicated `StateRead` group. There are no additions/removals or any other changes to the asm.

Should be *actually* ready for a proper review now!

File: crates/check/src/lib.rs [hunk 3]
```diff
@@ -23,22 +21,16 @@
 //!
 //! - [`solution::check_predicates`] validates a solution's data against their associated predicates.
 //! - [`solution::check_predicate`] validates a single solution data against an associated predicate.
-//! - [`solution::check_predicate_constraints`] the predicate constraint checking part of solution
-//!   data validation.
-//! - [`solution::check_decision_variable_lengths`] checks the expected number of
-//!   decision variables.
```
Comment (supiket) [inline crates/check/src/lib.rs:29]: Same question as above

File: crates/vm/Cargo.toml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 [package]
-name = "essential-constraint-vm"
-version = "0.5.0"
-description = "The Essential constraint checking VM"
+name = "essential-vm"
+version = "0.6.0"
```
Comment (supiket) [inline crates/vm/Cargo.toml:3]: or 0.1.0?

File: crates/vm/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    ControlFlow(asm::TotalControlFlow),
+    /// `[asm::Crypto]` operations.
+    Crypto(asm::Crypto),
+    /// `[asm::Memory]` operations.
+    Memory(asm::Memory),
+    /// `[asm::Pred]` operations.
+    Pred(asm::Pred),
+    /// `[asm::Stack]` operations.
+    Stack(asm::Stack),
+}
```
Comment (supiket) [inline crates/vm/src/lib.rs:121]: Would it be useful to mention in the high level docs, or anywhere else you see fit, that the all ops = async (state read) ops + sync (all else) ops?

File: crates/vm/src/pred.rs [hunk 6]
```diff
@@ -0,0 +1,240 @@
+use std::collections::HashSet;
+
```
Comment (supiket) [inline crates/vm/src/pred.rs:2]: Will require `cargo fmt`
```suggestion
```

File: crates/vm/src/pred.rs [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+use essential_types::Word;
+
```
Comment (supiket) [inline crates/vm/src/pred.rs:4]: Will require `cargo fmt`
```suggestion
```

**Review by supiket — COMMENTED** [review]
This is huge 🤯 The asm changes on top of the previous changes look amazing. Feel free to leave my comments to followup PRs as you wish, as this PR is quite a lot to digest already.

File: crates/check/Cargo.toml [hunk 8]
```diff
@@ -9,12 +9,10 @@ license.workspace = true
 repository.workspace = true
 
 [dependencies]
-essential-constraint-vm = { workspace = true }
 essential-hash = { workspace = true, optional = true }
 essential-sign = { workspace = true }
-essential-state-read-vm = { workspace = true }
 essential-types = { workspace = true }
-rayon = { workspace = true }
```
[the remark this answers is not in the document]
Comment (nfurfaro) [reply crates/check/Cargo.toml:17]: wow, that's a big dep dropped.

**Review by nfurfaro — APPROVED** [review]
Nicely done!

File: crates/asm-spec/asm.yml [hunk 9]
```diff
@@ -1,565 +1,563 @@
...
+          short: PEX
+          description: |
+            Check if a solution to a predicate exists within the same solution
+            with the hash of the arguments and address.
+
+            Returns `true` if the predicate exists.
+          stack_in: [sha256(arg0len, arg0, argNlen, argN, contract_addr, predicate_addr)]
+          stack_out: [bool]
+
+    # Byte 4 reserved for Access ops
```
[the remark this answers is not in the document]
Comment (mitchmindtree) [reply crates/asm-spec/asm.yml:362]: Besides the de-nesting there are no changes to the opcodes in this PR, but doing doing a follow-up ASM opcode review PR sounds good 👌

File: crates/check/Cargo.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (mitchmindtree) [reply crates/check/Cargo.toml:17]: > Was this obsolete before this PR already?

Yee, see the comment in [this commit](https://github.com/essential-contributions/essential-base/pull/229/commits/49a923e960f40ffe0dd5907f32ba0a7472970f73)

File: crates/check/src/lib.rs [hunk 10]
```diff
@@ -10,8 +10,6 @@
 //! - [`predicate::check_signed_contract`] validates a signed contract.
 //! - [`predicate::check_contract`] validates a contract.
 //! - [`predicate::check`] validate an individual predicate.
-//! - [`predicate::check_state_reads`] validate a predicate's state read bytecode.
-//! - [`predicate::check_constraints`] validate a predicate's constraint bytecode.
```
[the remark this answers is not in the document]
Comment (mitchmindtree) [reply crates/check/src/lib.rs:14]: Yes exactly, the fns they point to no longer exist - we should probably have a `cargo doc` lint job of some sort to check for this.

File: crates/vm/Cargo.toml (hunk 4 above)
Comment (mitchmindtree) [reply crates/vm/Cargo.toml:3]: Yeah I wasn't sure which way to go on this 🤔 I think the version `0.6.0` is unchanged from the previous `essential-state-read-vm` version that `develop` is on. Because this is a rename, I think it might be a bit confusing to reset the version entirely? Maybe we should pick the first version that follows the last version of either of the VMs?

File: crates/asm/src/lib.rs [hunk 11]
```diff
@@ -190,27 +193,18 @@ mod tests {
...
-            Stack::Swap.into(),
-            Stack::Dup.into(),
+            Memory::Load.into(),
+            Memory::Store.into(),
         ];
         roundtrip(ops);
     }
 
     #[test]
+    #[allow(clippy::useless_conversion)]
```
[the remark this answers is not in the document]
Comment (mitchmindtree) [reply crates/asm/src/lib.rs:203]: Good spot! It looks like a couple tests got jumbled here 🧐 will fix

File: crates/asm-spec/asm.yml [hunk 12]
```diff
@@ -1,565 +1,563 @@
...
+        ShrI:
+          opcode: 0x27
+          description: Arithmetic shift right by number of bits.
+          panics:
+            - Number of bits is negative.
+            - Number of bits is greater than a Word.
+          stack_in: [lhs, rhs]
+          stack_out: ["lhs >> rhs"]
+
+    Access:
```
[the remark this answers is not in the document]
Comment (mitchmindtree) [reply crates/asm-spec/asm.yml:283]: Issue added: https://github.com/essential-contributions/essential-base/issues/232

[the remark this answers is not in the document]
Comment (supiket) [reply crates/asm-spec/asm.yml:283]: Opened #233

File: crates/asm-spec/asm.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (supiket) [reply crates/asm-spec/asm.yml:362]: #233

File: crates/vm/Cargo.toml (hunk 4 above)
Comment (supiket) [reply crates/vm/Cargo.toml:3]: Yeah that makes sense

File: crates/vm/src/lib.rs (hunk 5 above)
Comment (supiket) [reply crates/vm/src/lib.rs:121]: Oh yeah, I meant in this VM implementation. Passing this is fine though!

**Review by supiket — APPROVED** [review]
📦
